File: data_structures_and_algorithms/alvin-zablan/dynamic-programming/part04_howsum/javascript/howSum.js
/*
  Given a target sum and an array of numbers, determine what combination (if any) of numbers in the array 
  can lead to the target sum when added together? Return an array of the numbers that can be added together to 
  give the target sum. 

  This tree can be represented in the same way as the canSum problem. Starting with a node of the target sum, subract each 
  value in the numbers array from the node(as a branch). Recursively.

  Note that as the targetSum grows, the ability to compute this value diminishes exponentially. To combat that, we use memoisation.
*/

const howSum = (targetSum, numbers, memo = {}) => {
  //base cases
  if (targetSum in memo) return memo[targetSum];
  if (targetSum === 0) return []; //return empty array
  if (targetSum < 0) return null; //return null

  for (let num of numbers) {
    const remainder = targetSum - num;
    const remainderResult = howSum(remainder, numbers, memo);
    if (remainderResult !== null) {
      //add the remainder to the return array
      memo[targetSum] = [...remainderResult, num];
      return memo[targetSum];
    }
  }
  memo[targetSum] = null;
  return memo[targetSum];
};

// m = target sum
// n = numbers length

// Brute force space and time complexity
// time = O(m^n * m)
// space = O(m)

// Memoised space and time complexity
// time = O(n * m^2)
// space = O(m^2)

console.log(howSum(7, [2, 3]));
console.log(howSum(7, [5, 3, 4, 7]));
console.log(howSum(7, [2, 4]));
console.log(howSum(8, [2, 3, 5]));
console.log(howSum(300, [7, 14]));
console.log(howSum(0, [7, 14, -11]));
console.log(howSum(230, [7, 14, -11]));

File: data_structures_and_algorithms/alvin-zablan/dynamic-programming/part04_howsum/java/HowSum.java
import java.util.ArrayList;

public class HowSum{
  public static void main(String[] args) {

    System.out.println(doHowSum(7, new int[]{2,3}));      //true
    System.out.println(doHowSum(7, new int[]{5,3,4,7}));  //true
    System.out.println(doHowSum(7, new int[]{2,4}));      //false
    System.out.println(doHowSum(8, new int[]{2,3,5}));    //true
    //System.out.println(doHowSum(300, new int[]{7,14}));   //false
    
  }

  public static ArrayList<Integer> doHowSum(int targetSum, int[] numbers ){
    if(targetSum == 0) return new ArrayList<Integer>();
    if(targetSum < 0) return null;

    for(int num : numbers){
      int remainder = targetSum - num;
      ArrayList<Integer> remainderResult = doHowSum(remainder, numbers );
      if(remainderResult != null){
        remainderResult.add(num);
        return remainderResult;
      }
    }
    return null;
  }
}